src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,613 @@
+  }
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+}
+
+// Cached: false, Callback: thread
+TEST_F(ServiceControlClientImplQuotaTest, TestNonCachedQuotaThread) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(Invoke(&mock_quota_transport_,
+                             &MockQuotaTransport::AllocateQuotaUsingThread));
+
+  Status done_status;
+  AllocateQuotaResponse quota_response;
+
+  // Set the check status and response to be used in the on_quota_done
+  mock_quota_transport_.done_status_ = done_status;
+  mock_quota_transport_.quota_response_ = &quota_response;
+
+  StatusPromise status_promise;
+  StatusFuture status_future = status_promise.get_future();
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&status_promise, &done_status](Status status) {
+                   StatusPromise moved_promise(std::move(status_promise));
+                   moved_promise.set_value(status);
+                 });
+
+  // since it is a cache miss, transport should be called.
+  EXPECT_TRUE(MessageDifferencer::Equals(mock_quota_transport_.quota_request_,
+                                         quota_request1_));
+
+  status_future.wait();
+  EXPECT_EQ(status_future.get(), Status::OK);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+}
+
+// Cached: true, Callback: thread
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestCachedQuotaThreadMultipleRequests) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(Invoke(&mock_quota_transport_,
+                             &MockQuotaTransport::AllocateQuotaUsingThread));
+
+  Status done_status;
+  AllocateQuotaResponse quota_response;
+
+  // Set the check status and response to be used in the on_quota_done
+  mock_quota_transport_.done_status_ = done_status;
+  mock_quota_transport_.quota_response_ = &quota_response;
+
+  StatusPromise status_promise;
+  StatusFuture status_future = status_promise.get_future();
+  client_->Quota(quota_request1_, &quota_response,
+                 [&status_promise, &done_status](Status status) {
+                   StatusPromise moved_promise(std::move(status_promise));
+                   moved_promise.set_value(status);
+                 });
+  EXPECT_TRUE(MessageDifferencer::Equals(mock_quota_transport_.quota_request_,
+                                         quota_request1_));
+  status_future.wait();
+  EXPECT_EQ(status_future.get(), Status::OK);
+
+  for (int i = 0; i < 10; i++) {
+    StatusPromise status_promise;
+    StatusFuture status_future = status_promise.get_future();
```

Comment:
Not need to use future/promise,  it will be cache hit,  status should be set right there.

This can be called even before the first Quota call wait. (before line 586)